rest-api/test/unit_test/api_test/participant_counts_over_time_api_test.py

```diff
@@ -1,53 +1,601 @@
-                     organizationType=OrganizationType.UNSET))
+    self.hpo_dao = HPODao()
+
+    # Needed by ParticipantCountsOverTimeApi
+    self.hpo_dao.insert(HPO(hpoId=TEST_HPO_ID, name=TEST_HPO_NAME, displayName='Test',
+                       organizationType=OrganizationType.UNSET))
+
+    self.time1 = datetime.datetime(2017, 12, 31)
+    self.time2 = datetime.datetime(2018, 1, 1)
+    self.time3 = datetime.datetime(2018, 1, 2)
+    self.time4 = datetime.datetime(2018, 1, 3)
 
-    curr_date = datetime.date(2017, 12, 22)
     # Insert 2 weeks of dates
-    for _ in xrange(0, 14 * 1):
+    curr_date = datetime.date(2017, 12, 22)
+    for _ in xrange(0, 14):
       calendar_day = Calendar(day=curr_date )
       CalendarDao().insert(calendar_day)
       curr_date = curr_date + datetime.timedelta(days=1)
 
+  def _insert(self, participant, first_name=None, last_name=None, hpo_name=None,
+              time_int=None, time_mem=None, time_fp=None):
+    """
+    Create a participant in a transient test database.
+
+    :param participant: Participant object
+    :param first_name: First name
+    :param last_name: Last name
+    :param hpo_name: HPO name (one of PITT or AZ_TUCSON)
+    :param time_int: Time that participant fulfilled INTERESTED criteria
+    :param time_mem: Time that participant fulfilled MEMBER criteria
+    :param time_fp: Time that participant fulfilled FULL_PARTICIPANT criteria
+    :return: Participant object
+    """
+
+    if time_mem is None:
+      enrollment_status = EnrollmentStatus.INTERESTED
+    elif time_fp is None:
+      enrollment_status = EnrollmentStatus.MEMBER
+    else:
+      enrollment_status = EnrollmentStatus.FULL_PARTICIPANT
+
+    with FakeClock(time_int):
+      self.dao.insert(participant)
+
+    participant.providerLink = make_primary_provider_link_for_name(hpo_name)
+    with FakeClock(time_mem):
+      self.dao.update(participant)
+    summary = self.participant_summary(participant)
+
+    if first_name:
+      summary.firstName = first_name
+    if last_name:
+      summary.lastName = last_name
+
+    summary.dateOfBirth = datetime.date(1978, 10, 10)
+
+    summary.enrollmentStatus = enrollment_status
+    summary.hpoId = self.hpo_dao.get_by_name(hpo_name).hpoId
+
+    if time_mem is not None:
+      with FakeClock(time_mem):
+        summary.consentForElectronicHealthRecordsTime = time_mem
+
+    if time_fp is not None:
+      with FakeClock(time_fp):
+        summary.consentForElectronicHealthRecordsTime = time_fp
+        summary.questionnaireOnTheBasicsTime = time_fp
+        summary.questionnaireOnLifestyleTime = time_fp
+        summary.questionnaireOnOverallHealthTime = time_fp
+        summary.physicalMeasurementsTime = time_fp
+        summary.sampleStatus1ED04Time = time_fp
+        summary.sampleStatus1SALTime = time_fp
+
+    self.ps_dao.insert(summary)
+
+    return summary
+
   def test_get_counts_with_default_parameters(self):
```

Comment:
At some point you may want to add some tests for invalid inputs (e.g date ranges too big, start date after end date, missing parameters, etc.) just to make sure you get back the right error code.

Reply:
There are a few tests for invalid inputs (e.g. date ranges to big) in this existing code.  I've added TODOs for the particular untested instances of invalid parameters you've mentioned here.